### PR TITLE
`Exam mode`: Fix breadcrumbs when assessing exam exercises

### PR DESCRIPTION
**Pull request by Strohgelaender**
#### General
- [ ] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [ ] This is a small issue that I tested locally and was confirmed by another developer on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.cit.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/development-process/#naming-conventions-for-github-pull-requests).
#### Client
- [x] **Important**: I implemented the changes with a very good performance, prevented too many (unnecessary) REST calls and made sure the UI is responsive, even with large data.
- [x] I followed the [coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client/).
- [x] Following the [theming guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-design/), I specified colors only in the theming variable files and checked that the changes look consistent in both the light and the dark theme.
- [ ] I added multiple integration tests (Jest) related to the features (with a high test coverage), while following the [test guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-tests/).
- [x] I added `authorities` to all new routes and checked the course groups for displaying navigation elements (links, buttons).
- [x] I documented the TypeScript code using JSDoc style.
- [ ] I added multiple screenshots/screencasts of my UI changes.
- [ ] I translated all newly inserted strings into English and German.

### Motivation and Context
During the exam mode testing we noticed that opening the exercise assessment dashboard loses the exam context. This means that navigating back into the exam using the breadcrumbs no longer works. You instead get into the course assessment dashboard, which looks similar but does not contain the exam exercises, which can be confusing. 

### Description
We now keep the examId in the urls while navigating to the assessment dashboard and when starting an assessment.

While the components support this already, some of the necessary routes were missing.

### Steps for Testing
Prerequisites:
- 1 Tutor
- 1 Exam with exercises of each type

1. Open the exercise assessment dashboard as a tutor (exams -> assessment -> exercise dashboard)
2. From now on, check the breadcrumbs with every button press and make sure the exam context gets maintained.
3. Start new assessments, open existing ones, etc. (do typical assessment tasks).

#### Performance Review
- [x] I confirm that the client changes (in particular related to REST calls and UI responsiveness) are implemented with a very good performance 
- [x] I confirm that the server changes (in particular related to database calls) are implemented with a very good performance
#### Code Review
- [x] Code Review 1
- [x] Code Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2

### CODE REVIEW

**Review by rriyaldhi — APPROVED** [review]
Tested on TS4 and works as described.

**Review by laadvo — APPROVED** [review]
Tested on ts4, working as expected :+1:

File: src/main/webapp/app/exam/manage/exam-management.route.ts [hunk 1]
```diff
@@ -814,14 +805,23 @@ export const examManagementRoute: Routes = [
         canActivate: [UserRouteAccessService],
     },
     {
-        path: ':examId/exercise-groups/:exerciseGroupId/modeling-exercises/:exerciseId/submissions/:submissionId/assessments/:resultId',
```
Comment (JohannesStoehr) [inline src/main/webapp/app/exam/manage/exam-management.route.ts:817]: The navigation.utils.ts add the `s/resultId` in lines 132-135. Did you check if this change has an influence on the navigation there?

Comment (Strohgelaender) [reply src/main/webapp/app/exam/manage/exam-management.route.ts:817]: Yes. This was actually duplicated in the routing and still exists in line 853.

File: src/main/webapp/app/exercises/shared/dashboards/tutor/exercise-assessment-dashboard.component.ts [hunk 2]
```diff
@@ -678,10 +678,10 @@ export class ExerciseAssessmentDashboardComponent implements OnInit {
     getAssessmentLink(submission: Submission | 'new'): string[] {
         const submissionUrlParameter: number | 'new' = submission === 'new' ? 'new' : submission.id!;
         let participationId = undefined;
-        if (submission !== 'new' && submission.participation !== undefined) {
-            participationId = submission.participation!.id;
+        if (submission !== 'new' && submission.participation) {
+            participationId = submission.participation.id;
         }
-        return getLinkToSubmissionAssessment(this.exercise.type!, this.courseId!, this.exerciseId, participationId, submissionUrlParameter, this.examId, this.exerciseGroupId);
+        return getLinkToSubmissionAssessment(this.exercise.type!, this.courseId, this.exerciseId, participationId, submissionUrlParameter, this.examId, this.exerciseGroupId);
```
Comment (pal03377) [inline src/main/webapp/app/exercises/shared/dashboards/tutor/exercise-assessment-dashboard.component.ts:684]: Refactoring

**Review by pal03377 — APPROVED** [review]
Code changes look good ✅  As far as I see, the only actual change happens in `exam-management.route.ts` - the rest is only refactoring (which would have been helpful to know when reviewing the code, a small comment in the PR description would be nice next time :) ).

Well done!

File: src/main/webapp/app/exam/manage/exam-management.route.ts (hunk 1 above)
Comment (Strohgelaender) [reply src/main/webapp/app/exam/manage/exam-management.route.ts:817]: I looked at the course routes where the same authorities get used, so this should be fine. I also think it makes more sense that way, considering that tutors should not assess a specific result (double blind).
grafik

File: src/main/webapp/app/exercises/shared/dashboards/tutor/exercise-assessment-dashboard.component.ts [hunk 3]
```diff
@@ -203,7 +203,6 @@ export class ExerciseAssessmentDashboardComponent implements OnInit {
 
         if (this.route.snapshot.paramMap.has('examId')) {
             this.examId = Number(this.route.snapshot.paramMap.get('examId'));
-            this.exerciseGroupId = Number(this.route.snapshot.paramMap.get('exerciseGroupId'));
```
[the remark this answers is not in the document]
Comment (Strohgelaender) [reply src/main/webapp/app/exercises/shared/dashboards/tutor/exercise-assessment-dashboard.component.ts:206]: The URL used for the assessment dashboard is `/course-management/courseId/exams/examId/assessment-dashboard/exerciseId' (see naviation.utils line 142). This line expects another `exerciseGroupId` as a parameter that simply does not exist. I removed this line (which sets it to always 0) and replaced it below by the exercise group id retrieved from the exercise.

(Sorry, I wanted to wrote this into the description but forgot)

File: src/main/webapp/app/exam/manage/exam-management.route.ts [hunk 4]
```diff
@@ -814,14 +805,23 @@ export const examManagementRoute: Routes = [
...
     },
+    {
+        path: ':examId/exercise-groups/:exerciseGroupId/programming-exercises/:exerciseId/submissions/:submissionId/assessment',
+        component: !isOrion ? CodeEditorTutorAssessmentContainerComponent : OrionTutorAssessmentComponent,
+        data: {
+            authorities: [Authority.ADMIN, Authority.INSTRUCTOR, Authority.EDITOR, Authority.TA],
+            pageTitle: 'artemisApp.programmingExercise.home.title',
+        },
+        canActivate: [UserRouteAccessService],
+    },
```
[the remark this answers is not in the document]
Comment (Strohgelaender) [reply src/main/webapp/app/exam/manage/exam-management.route.ts:824]: The orion component is just a small wrapper around the `CodeEditorTutorAssessmentContainerComponent`, it's done the same way for course exercises. So this should be fine, however I could not test it.

[the remark this answers is not in the document]
Comment (pal03377) [reply src/main/webapp/app/exam/manage/exam-management.route.ts:824]: I messaged @Kroko-fant on Slack about this and asked him whether he could test this. As far as I'm concerned, this does not have to happen right away though.

**Review by DominikRemo — APPROVED** [review]
Code LGTM

**Pull request merged**